Async seek (#1435)
bevyengine/bevy-website: release-content/0.14/release-notes/12547_Add_AsyncSeek_trait_to_Reader_to_be_able_to_seek_inside_as.md

<!-- Add AsyncSeek trait to Reader to be able to seek inside asset loaders -->
<!-- https://github.com/bevyengine/bevy/pull/12547 -->

> When loading assets you might be dealing with big data.

Reviewer: ```suggestion
Assets can be huge, and you don't always need all of the data contained in a single file.
```

> In Bevy you're able to add your [`own asset loaders`]. Now you will be able to seek to an offset of your choice on load using async.

Reviewer: ```suggestion
Bevy allows you to add your [own asset loaders].
Starting in Bevy 0.14,  you can now seek to an offset of your choice, reading partway through the file.
```

```rust
#[derive(Default)]
struct UniverseLoader;

#[derive(Asset, TypePath, Debug)]
struct JustALilAsteroid([u8; 128]); // Each lil' asteroid uses this much data

impl AssetLoader for UniverseLoader {
    type Asset = JustALilAsteroid;
    type Settings = ();
    type Error = std::io::Error;
    async fn load<'a>(
        &'a self,
        reader: &'a mut Reader<'_>,
        _settings: &'a (),
        _load_context: &'a mut LoadContext<'_>,
    ) -> Result<JustALilAsteroid, Self::Error> {
        // The universe is big, and our lil' asteroids don't appear until this offset
        // in the celestial file format!
        let offset_of_lil_asteroids = 5_000_000_000_000;

        // Skip vast parts of the universe with the new async seek trait!
        reader
            .seek(SeekFrom::Start(offset_of_lil_asteroids))
            .await?;

        let mut asteroid_buf = [0; 128];
        reader.read_exact(&mut asteroid_buf).await?;

        Ok(JustALilAsteroid(asteroid_buf))
    }

    fn extensions(&self) -> &[&str] {
        &["celestial"]
    }
}
```

> This works by the fact that Bevy's [`reader`] type passed into the asset loader's `load` function now implements [`AsyncSeek`].

Reviewer: ```suggestion
This works because Bevy's [`reader`] type passed into the asset loader's `load` function now implements [`AsyncSeek`].
```
